Append version to dist URL to force re-download

mcaskill opened the pull request:
## Checklist:

- [x] I've read the [Contributing page](https://github.com/junaidbhura/composer-wp-pro-plugins/blob/master/CONTRIBUTING.md).
- [x] I've created an issue and referenced it here.
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [ ] My code has proper inline documentation.

## Description

Currently, this Composer plugin reuses the same cache key regardless of the specified package version because the package's `dist.url` does not change between version bumps.

```json
"version": "5.9.0",
"dist": {
    "type": "zip",
    "url": "https://www.advancedcustomfields.com"
}
```

Based on Composer's [documentation](https://getcomposer.org/doc/05-repositories.md#package-2) for packages that do not support Composer (and this [snippet from the source code](https://github.com/composer/composer/blob/1.10.10/src/Composer/Downloader/FileDownloader.php#L276-L285)) the `dist.url` is expected to reflect the version:

```json
"version": "5.9.0",
"dist": {
    "type": "zip",
    "url": "https://github.com/AdvancedCustomFields/acf/archive/5.9.0.zip"
}
```

This means if you were to change the `dist.url` (any random URI) along with a version bump, you'd see that Composer would use a new cache key with that new version.

Since the real archive files for these premium WordPress plugins are hidden behind complex URLs, API response keys, and often expire, this Composer plugin needs to assign some kind of unique stand-in for the `dist.url` to serve as a unique cache key for each version.

## How has this been tested?

I've tested this on a client project that uses Advanced Custom Fields Pro, Gravity Forms, and Polylang Pro.

## Types of changes

This pull request introduces a change that when a download URL is resolved, the package's unique name (name + normlized version) is appended to the internal `dist.url` (as a hashed URI fragment) in order to generate a new cache key to trigger the download of the package update.

Similar to [ffraenz/private-composer-installer](https://github.com/ffraenz/private-composer-installer/blob/v4.0.0/src/PrivateComposerInstaller/Plugin.php#L75-L85).

Resolves #7, #18

mcaskill committed:
Append version to dist URL to force re-download

This allows Composer to cache the updated version under a new hash.

Resolves #7, #18

mcaskill commented on Installer.php:
```diff
@@ -79,33 +79,41 @@ public function getDownloadUrl( PackageEvent $event ) {
+			if ( false !== strpos( $package_url, '%version%' ) ) {
+				$package->setDistUrl( str_replace( '%version%', $package_version, $package_url ) );
+			} elseif ( false === strpos( $package_url, $package_version ) ) {
+                $package->setDistUrl( $package_url . '#v' . $package_version );
```
Given how generic the initial `dist.url` between WP plugin packages and how certain classes from `WPProPlugins\Plugins\*` support add-ons, we end up with dist URLs and cache keys such as:

- `junaidbhura/gravityforms`:
  - Dist URL: `https://www.gravityforms.com/#v2.4.20.3`
  - Cache Path: `junaidbhura/gravityforms/335917ed5650d888fcf4ad07a379d42d76e43b7b.zip`
- `junaidbhura/gravityformsmailchimp`:
  - Dist URL: `https://www.gravityforms.com/#v4.7.1`
  - Cache Path: `junaidbhura/gravityformsmailchimp/67b72a2a27e2aed83e0af5d338b5799fa73045c2.zip`

If you eventually support a private WP plugin where add-ons share the same version as the main package (like Symfony does between all its components), then we will end up with the same cache key across packages. They ZIPs are stored in different directories, nonetheless this could be problematic.

It may be a better choice to use `PackageInterface::getUniqueName()` (includes vendor, name, and version) instead of `PackageInterface::getPrettyVersion()` (just the version).

mcaskill replied:
Alternatively, we could delegate the filtering of the `dist.url` to the plugin classes themselves:

```php
class AcfPro {
    // ...
	/**
	 * Get the dist URL for this plugin.
	 *
	 * @return string
	 */
	public function getDistUrl() {
		return 'https://www.advancedcustomfields.com/#v' . $this->version;
	}
    // ...
}
```

This would require some more significant changes to this library but may be easier to maintain long-term.

junaidbhura replied:
Hey @mcaskill - first of all - thanks for finding the problem! 🎉 

Do you think we could solve the problem by using a hash instead of the version number? So something like `https://www.gravityforms.com/#v<base64_encode( $package_name . $package_version )>` . That way even if the addons have the same versions, there would be no conflict?

mcaskill replied:
Yeah, we could hash the unique name. Should be okay.

mcaskill replied:
Do you want the hashing to occur in the current (198bdf8e) proposal or in the [delegated dist URL](https://github.com/junaidbhura/composer-wp-pro-plugins/pull/20#discussion_r484113895) concept?

junaidbhura replied:
Hey @mcaskill lets do it with the current proposal for now! Thanks!

mcaskill committed:
Improve locked dist URL to minimize conflicts

Instead of appending the package's pretty version to the initial dist URL, we are appending a SHA1 hash of the package's unique name (`<name>-<version>`).

This will prevent duplicate cache keys between a plugin and its add-ons that may share the same version.

Removed:
- Variable `$package_version` since usage is limited to `switch` statement.
- Placeholder `%version%` since its an unused concept and confusing with usage of unique name instead of only the version.

junaidbhura committed:
Update indentation and add comment
